File: cpp/examples/group_gain.cpp
// File: group_gain.cpp
// Project: examples
// Created Date: 14/09/2023
// Author: Shun Suzuki
// -----
// Last Modified: 15/09/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
//

#include "autd3.hpp"
#include "autd3/link/debug.hpp"
#include "runner.hpp"
#include "util.hpp"

int main() try {
  auto autd = autd3::Controller::builder().add_device(autd3::AUTD3(autd3::Vector3::Zero(), autd3::Vector3::Zero())).open_with(autd3::link::Debug());

  const auto cx = autd.geometry().center().x();
  autd3::gain::Focus g1(autd.geometry().center() + autd3::Vector3(0, 0, 150));
  autd3::gain::Null g2;

  const auto g = autd3::gain::Group([&cx](const autd3::Device& dev, const autd3::Transducer& tr) -> std::optional<const char*> {
                   if (tr.position().x() < cx) return "focus";
                   return "null";
                 })
                     .set("focus", g1)
                     .set("null", g2);

  autd3::modulation::Sine m(150);
  autd.send(m, g);

  autd.close();

  return 0;

} catch (std::exception& e) {
  print_err(e);
  return -1;
}

File: src/examples/src/group_gain.rs
/*
 * File: group.rs
 * Project: src
 * Created Date: 02/09/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 06/10/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 *
 */

use anyhow::Result;

use autd3::prelude::*;

fn main() -> Result<()> {
    let mut autd = Controller::builder()
        .add_device(AUTD3::new(Vector3::zeros(), Vector3::zeros()))
        .open_with(Nop::builder())?;

    let cx = autd.geometry().center().x;
    let g1 = Focus::new(autd.geometry()[0].center() + Vector3::new(0., 0., 150.0 * MILLIMETER));
    let g2 = Null::new();
    let g = Group::new(move |_dev, tr: &LegacyTransducer| {
        if tr.position().x < cx {
            Some("focus")
        } else {
            Some("null")
        }
    })
    .set("focus", g1)
    .set("null", g2);

    let m = Sine::new(150);
    autd.send((m, g))?;

    autd.close()?;

    Ok(())
}
